### PR TITLE
feat: Add Network module in Core2

### PR DESCRIPTION
**Pull request by KevinBoulongne**

### CODE REVIEW

File: Core2/Matomo/proguard-rules.pro
Comment (LunarX) [inline Core2/Matomo/proguard-rules.pro:1]: Why remove the last new line at the end of the file?

File: Core2/Network/proguard-rules.pro
Comment (LunarX) [inline Core2/Network/proguard-rules.pro:1]: In the same idea, why no new line at the end of the file?

File: Core2/Network/consumer-rules.pro
Comment (LunarX) [inline Core2/Network/consumer-rules.pro:1]: Is this required if it's empty?

File: Core2/Network/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+android {
+    namespace = "com.infomaniak.network"
+    compileSdk = 34
+
+    defaultConfig {
+        minSdk = 24
+
+        testInstrumentationRunner = "androidx.test.runner.AndroidJUnitRunner"
```
Comment (LunarX) [inline Core2/Network/build.gradle.kts:13]: Do you need tests if there's no tests in your module?

Comment (LunarX) [inline Core2/Network/build.gradle.kts:13]: Do you need tests if there's no tests in your module?

**Review by LunarX — CHANGES_REQUESTED** [review]

File: Core2/Matomo/proguard-rules.pro
Comment (KevinBoulongne) [reply Core2/Matomo/proguard-rules.pro:1]: It was the only one with it, I just uniformized.

File: Core2/Network/proguard-rules.pro
Comment (KevinBoulongne) [reply Core2/Network/proguard-rules.pro:1]: Automatically created by Android Studio.

File: Core2/Network/consumer-rules.pro
Comment (KevinBoulongne) [reply Core2/Network/consumer-rules.pro:1]: Automatically created by Android Studio, we have it for every modules.

File: Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    val isNetworkAvailable: Flow<Boolean> = callbackFlow {
+        val networks = mutableListOf<Network>()
+
+        val callback = object : NetworkCallback() {
+
+            override fun onAvailable(network: Network) {
+                launch {
+                    mutex.withLock {
+                        networks.add(network)
+                        sendNetworkAvailability(networks)
```
Comment (sirambd) [inline Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt:52]:
```suggestion
                        send(element = hasAvailableNetwork(networks))
```

File: Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                    }
+                }
+            }
+
+            override fun onLost(network: Network) {
+                launch {
+                    mutex.withLock {
+                        networks.remove(network)
+                        sendNetworkAvailability(networks)
```
Comment (sirambd) [inline Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt:61]: send(element = hasAvailableNetwork(networks))

File: Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+                    }
+                }
+            }
+        }
+
+        launch(ioDispatcher) {
+            send(getInitialNetworkAvailability(connectivityManager))
+        }
+
+        registerNetworkCallback(connectivityManager, callback)
```
Comment (sirambd) [inline Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt:71]:
```suggestion
        registerNetworkCallback(connectivityManager, callback, send = ::send)
```

File: Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+        awaitClose { unregisterNetworkCallback(connectivityManager, callback) }
+    }
+
+    private suspend fun ProducerScope<Boolean>.registerNetworkCallback(
+        connectivityManager: ConnectivityManager,
+        callback: NetworkCallback,
+    ) {
```
Comment (sirambd) [inline Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt:79]:
```suggestion
    private suspend fun registerNetworkCallback(
        connectivityManager: ConnectivityManager,
        callback: NetworkCallback,
        send: suspend (Boolean) -> Unit,
    ) {
```

File: Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+    private fun unregisterNetworkCallback(connectivityManager: ConnectivityManager, callback: NetworkCallback) {
+        runCatching {
+            connectivityManager.unregisterNetworkCallback(callback)
+        }.onFailure { exception ->
+            Sentry.captureException(exception)
+        }
+    }
+
+    @Suppress("DEPRECATION")
```
Comment (sirambd) [inline Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt:99]: Useless

File: Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    private fun getInitialNetworkAvailability(connectivityManager: ConnectivityManager): Boolean {
+        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            connectivityManager.activeNetwork?.let(::hasInternetConnectivity) ?: false
+        } else {
+            connectivityManager.activeNetworkInfo?.isConnected ?: false
+        }
```
Comment (sirambd) [inline Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt:105]:
```suggestion
        return connectivityManager.activeNetwork?.let(::hasInternetConnectivity) ?: false
```

File: Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    private fun ProducerScope<Boolean>.sendNetworkAvailability(networks: List<Network>) {
+        launch(ioDispatcher) { send(hasAvailableNetwork(networks)) }
+    }
```
Comment (sirambd) [inline Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt:110]: Useless

File: Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        return NetworkRequest.Builder().apply {
+            addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
+        }.build()
+    }
+
+    private fun hasInternetConnectivity(network: Network) = runCatching {
+        network.getByName(ROOT_SERVER_URL) != null
+    }.getOrDefault(false)
+
+    private suspend fun hasAvailableNetwork(networks: List<Network>) = mutex.withLock { networks.any(::hasInternetConnectivity) }
```
Comment (sirambd) [inline Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt:122]:
```suggestion
    private suspend fun hasAvailableNetwork(networks: List<Network>) = withContext(ioDispatcher) {
        networks.any(::hasInternetConnectivity)
    }
```

File: Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        connectivityManager: ConnectivityManager,
+        callback: NetworkCallback,
+    ) {
+        runCatching {
+            connectivityManager.registerNetworkCallback(networkRequestBuilder(), callback)
+        }.onFailure { exception ->
+            // Fix potential Exception thrown by ConnectivityManager on Android 11
+            // Already fixed in Android S and above
+            // https://issuetracker.google.com/issues/175055271
+            Sentry.captureException(exception)
```
Comment (sirambd) [inline Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt:86]: SentryLog.e(TAG, "Android 11 exception", exception)

**Review by sirambd — CHANGES_REQUESTED** [review]

File: Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt (hunk 7 above)
Comment (KevinBoulongne) [reply Core2/Network/src/main/java/com/infomaniak/network/NetworkAvailability.kt:105]: This file will be used in the Core project, which is `coreMinSdk = 21`, so I'll keep this version check.

File: Core2/Network/build.gradle.kts [hunk 11]
```diff
@@ -0,0 +1,35 @@
+plugins {
+    alias(libs.plugins.android.library)
+    alias(libs.plugins.kotlin.android)
+}
+
+android {
+    namespace = "com.infomaniak.network"
+    compileSdk = 34
```
[the remark this answers is not in the document]
Comment (KevinBoulongne) [reply Core2/Network/build.gradle.kts:8]: Fixed in #212.

File: Core2/Network/build.gradle.kts [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+            isMinifyEnabled = false
+            proguardFiles(getDefaultProguardFile("proguard-android-optimize.txt"), "proguard-rules.pro")
+        }
+    }
+    compileOptions {
+        sourceCompatibility = JavaVersion.VERSION_1_8
+        targetCompatibility = JavaVersion.VERSION_1_8
+    }
+    kotlinOptions {
+        jvmTarget = "1.8"
```
[the remark this answers is not in the document]
Comment (KevinBoulongne) [reply Core2/Network/build.gradle.kts:28]: Fixed in #212.

File: Core2/Network/build.gradle.kts [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+    buildTypes {
+        release {
+            isMinifyEnabled = false
+            proguardFiles(getDefaultProguardFile("proguard-android-optimize.txt"), "proguard-rules.pro")
+        }
+    }
+    compileOptions {
+        sourceCompatibility = JavaVersion.VERSION_1_8
+        targetCompatibility = JavaVersion.VERSION_1_8
```
[the remark this answers is not in the document]
Comment (KevinBoulongne) [reply Core2/Network/build.gradle.kts:25]: Fixed in #212.

**Review by sirambd — APPROVED** [review]

**Review by LunarX — APPROVED** [review]

**Pull request merged**